packages/pwa-kit-react-sdk/src/ssr/universal/hooks/use-server-effect.tsx

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import React, {useContext, useState, useEffect} from 'react'
+import {useUID} from 'react-uid' // TODO: Probably not required
+import useGetPropsArgs, {GetPropsArgs} from './use-get-props-args'
+
+// Globals
+export const DEFAULT_CONTEXT_KEY = '__SERVER_EFFECTS__'
+const contexts = []
+const isServer = typeof window === 'undefined'
+
+// Type Definitions
+type ServerEffectState = {
+    data: object;
+    loading: boolean;
+    error: Error;
+}
+type DidUpdateFn = (serverEffectArgs: GetPropsArgs) => Promise<any>
```

Comment:
```suggestion
type DidUpdateFn<T> = (serverEffectArgs: GetPropsArgs) => Promise<T>
```
The returned data will be passed around, we want to preserve it!